## Module 11: Intro to Correlation and Linear Regression 

***

You've learned to explore your dataset features, clean up messy data, and use descriptive statistics to summarize the characteristics of your dataset. 

We are now moving forward with exploring the relationship <b>between</b> variables. Determining how your variables interact is one of the most useful skills you will learn while working with data. 

The first step is understanding the dependency of your variables. 

***

In [2]:
import pandas as pd
import numpy as np

## <font color=DODGERBLUE>Independent and Dependent Variables</font>

When you start to look at the relationship between variables, there are two classifications of variables that need to be considered. 

***

### Independent Variables (IV)
An independent variable, also known as a predictor variable, is a variable that is independent of the other variables in your dataset. Consider this variable the "cause". Independent variables are commonly represented with "x".  

### Dependent Variables (DV)
A dependent variable, also known as an outcome variable, is a variable whose value is dependent on the values of the other variables in your dataset. Consider this variable the "effect". We assume that changes in the values of the independent variable(s) will result in changes to the dependent variable. Dependent variables are commonly represented with "y". 

***

### Examples 
The distinction between IV and DV is an important concept to master because several analyses require identification of which variables are IV and DV. IV's cause a change in the DV, and it isn't possible for the DV to cause a change in the IV.

#### Do flowers grow fastest under fluorescent or natural light?

    - IV: Type of light flowers are grown under
    - DV: Rate of flower growth

#### What is the effect of diet and regular soda on blood sugar levels?

    - IV: Type of soda
    - DV: Blood sugar levels

#### How does cellphone use before bedtime influence sleep?

    - IV: Amount of phone use before bedtime
    - DV: Hours of sleep, quality of sleep, restfulness, etc. 

In [3]:
df = pd.read_csv("EduGradeData.csv")

df.head()

## which variables can be considered independent and dependent?

fname     lname  gender  age  exercise level_of_fit  hours  \
0   Marcia      Pugh  female   17         3          low     10   
1   Kadeem  Morrison    male   18         4          low      4   
2     Nash    Powell    male   18         5          low      9   
3  Noelani    Wagner  female   14         2         high      7   
4  Noelani    Cherry  female   18         4          low     15   

  level_of_study  grade home_state  
0       moderate   82.4         NJ  
1            low   78.2         MA  
2       moderate   79.3         OH  
3       moderate   83.2         FL  
4           high   87.4         OH

## <font color=GOLDENROD>Your Turn</font>

    1. Import the "insurance.csv" file; name the dataset 'ins'. Preview the first 5 rows. 
    2. In the space below, determine what type of variable is in each column (qualitative/quantitative). 
    3. What variable(s) could be considered dependent?

In [706]:
import pandas as pd
import numpy as np
ins = pd.read_csv('insurance.csv')
ins.head(5)

age     sex     bmi  children smoker drinker  comorbidities prescription  \
0   54  female  47.410         0    yes     yes              5          yes   
1   45    male  30.360         0    yes     yes              5          yes   
2   52    male  34.485         3    yes     yes              2          yes   
3   31  female  38.095         1    yes     yes              4          yes   
4   33  female  35.530         0    yes     yes              5          yes   

    provider_type      charges  
0  out-of-network  63770.42801  
1  out-of-network  62592.87309  
2  out-of-network  60021.39897  
3  out-of-network  58571.07448  
4  out-of-network  55135.40209

In [4]:

#sex,smoker,drinker,prescription and provider_type are qualitative variables

#age,childern,bmi,comorbidities and charges are quantittive variables

# charges is the dependent variable

# <font color=DODGERBLUE>Introduction to Correlation & Association</font>

***

### Correlation
***
Correlation is a tool that can be used to determine th strength of a relationship between two numeric variables. Correlation describes the direction (+/-) and magnitude (how large) of a <b>linear relationship</b> exists between two numeric variables. This is the initial check if your numeric variables have any kind of meaningful relationship. 

Correlation values range from "-1" to "+1". Variables that are positively correlated are closer to "+1" and variables that are negatively correlated are closer to "-1". 

#### Positively correlated variables move in the same direction - as one variable increases, the other variable increases in the same direction.

        - Increase in study time >> increases in test score
        - Decrease in sugar consumption >> decrease in blood sugar levels
        
#### Negatively correlated variables move in the opposite direction - as one variable increases, the other variable decreases (or vice versa). 

        - Decrease in daily spending >> increase in total savings
        - Increase in weight >> decrease in mobility

In [7]:
df.head()

fname     lname  gender  age  exercise level_of_fit  hours  \
0   Marcia      Pugh  female   17         3          low     10   
1   Kadeem  Morrison    male   18         4          low      4   
2     Nash    Powell    male   18         5          low      9   
3  Noelani    Wagner  female   14         2         high      7   
4  Noelani    Cherry  female   18         4          low     15   

  level_of_study  grade home_state  
0       moderate   82.4         NJ  
1            low   78.2         MA  
2       moderate   79.3         OH  
3       moderate   83.2         FL  
4           high   87.4         OH

In [8]:
## Create a correlation matrix

df.corr()

age  exercise     hours     grade
age       1.000000 -0.003643 -0.017467 -0.007580
exercise -0.003643  1.000000  0.021105  0.161286
hours    -0.017467  0.021105  1.000000  0.801955
grade    -0.007580  0.161286  0.801955  1.000000

### Strength of Relationship between Variables
***
The strength of the relationship between variables is determined by the value of the correlation coefficient. To interpret its value, see which of the following values your correlation coefficient is closest to:

- <b>Exactly –1</b>: A perfect downhill (negative) linear relationship
- <b>–0.70</b>: A strong downhill (negative) linear relationship
- <b>–0.50</b>: A moderate downhill (negative) relationship
- <b>–0.30</b>: A weak downhill (negative) linear relationship
- <b>0</b>: No linear relationship
- <b>+0.30</b>: A weak uphill (positive) linear relationship
- <b>+0.50</b>: A moderate uphill (positive) relationship
- <b>+0.70</b>: A strong uphill (positive) linear relationship
- <b>Exactly +1</b>: A perfect uphill (positive) linear relationship

#### It is important to understand that <u>correlation is not the same as causation</u>. Identifying a correlation between two factors does not automatically mean one factor causes another factor to occur.

***

## <font color=GOLDENROD>Your Turn</font>

    1. Create a correlation matrix with the insurance dataset. 
    2. What can you say about the relationship between the independent variables and the dependent variable?

In [10]:
ins.corr()

age       bmi  children  comorbidities   charges
age            1.000000  0.109272  0.042469       0.611903  0.299008
bmi            0.109272  1.000000  0.012759       0.065460  0.198341
children       0.042469  0.012759  1.000000      -0.282273  0.067998
comorbidities  0.611903  0.065460 -0.282273       1.000000  0.505246
charges        0.299008  0.198341  0.067998       0.505246  1.000000

In [ ]:
# comorbidities has a moderate positive relationship with the charges but the rest have weak relationship with the charges

### Qualitative Variables and Association
***
The correlation matrix is great for examining the relationship between two numeric variables, however, this method does not work when you want to assess the relationship between qualitative variables, or one qualitative and one numeric variable. 

For these situations, we are interested in determining the differences between groups. These differences include variations in counts or means. If we identify notable differences between groups (the average test score for male students is 89, and the average test score for female students is 92) - this is evidence that there is some underlying relationship present that can be further explored. 

In [4]:
## Two qualitative variables

pd.crosstab(df["gender"], df["level_of_fit"])

level_of_fit  high  low
gender                 
female         387  613
male           410  590

In [5]:
## One qualitative variable, one numeric

df["grade"].groupby(df["gender"]).mean()

gender
female    82.7173
male      82.3948
Name: grade, dtype: float64

## <font color=GOLDENROD>Your Turn</font>

    1. Using the insurance dataset, determine if there is any association between the categorical independent variables and the dependent variable. 

In [12]:
ins['charges'].groupby(ins['sex']).mean()

sex
female    12569.578844
male      13956.751178
Name: charges, dtype: float64

In [13]:
ins['charges'].groupby(ins['smoker']).mean()

smoker
no      8434.268298
yes    32050.231832
Name: charges, dtype: float64

In [15]:
ins['charges'].groupby(ins['drinker']).mean()

drinker
no      7538.063708
yes    31357.615725
Name: charges, dtype: float64

In [17]:
ins['charges'].groupby(ins['children']).mean()

children
0    12365.975602
1    12731.171832
2    15073.563734
3    15355.318367
4    13850.656311
5     8786.035247
Name: charges, dtype: float64

In [18]:
ins['charges'].groupby(ins['provider_type']).mean()

provider_type
in-network         7084.809598
out-of-network    20855.774238
Name: charges, dtype: float64

### Variance and your Dependent Variable
***
Understanding which variables have a relationship with your dependent variable is the first step. The next step is to better understand what is influencing that relationship, or the variance seen in your dependent variable. <B>Variance</B> describes the variation in values for a specific variable. 

We know that some people are taller than others - not everyone is exactly the same height. If we were to ask 100 people their specific height - we would not receive 100 identical answers; instead, we would likely receive different values (with some overlap). This variation in values is variance. When variance is observed, the next step is to determine what the factors are between the variance - in other words, what factors are responsible for the difference in height between the 100 people? Gender is likely an important consideration - as males tend to be taller on average. But is that the only reason some people are taller or shorter than others? Maybe ethnicity is influencing this variance - some ethnic groups are naturally taller than others. Genetics might also contribute; as taller parents typically have taller children.

Gender, ethnicity, genetics, etc - are all factors that may contribute to the variation in height. In this example, these factors are independent variables and the height of the individual is the dependent variable. The key question is: <b>how do changes in the independent variables explain the variance in the dependent variable?</b> An additional question is: <b>how much of the variation in the dependent variable can be attributed to each independent variable?</b>

Once we understand the factors that are influencing the dependent variable, we can purposefully manipulate the values of the independent variables to predict how that will change the dependent variable. This is where regression comes in handy! 

In [ ]:
df.head(10)

# <font color=DODGERBLUE>Introduction to Linear Regression</font>
***

<b>Linear Regression</b> is a powerful statistical tool that allows for a closer examination of the linear relationship between <b>a continuous dependent variable</b> and various independent variables. 

Linear regression allows you to :
1. Determine if the relationship between the dependent and independent variables is statistically significant (or accurate).
2. Identify how much of the variation in the dependent variable is explained by the selection of independent variables. 
3. Determine the direction and magnitude of the relationships between variables, and 
4. Predict what the value of the dependent variable would be given specific input from the independent variables. 

You can use linear regression to predict the salary of a lawyer (DV) based on the number of years they practiced law (IV). You could also determine just how much of the variation in salary is attributed to the number of years they have practiced law.

***

A simple linear regression models the relationships between a single dependent and independent variable, <b>where the independent variable is predicting the value of the dependent variable</b>. A linear regression model is mathematically represented by the formula of a line:
### \begin{align}  y = mx + b \end{align}
Where “y” is the the value of the dependent variable, “m” is the slope (also known as the coefficient), “x” represents the value of the independent variable, and “b” is the y-intercept (also known as the constant) which is the value of “y” when the coefficient is equal to 0. 

<b>Linear regression models will determine the line-of-best-fit, also known as the regression line, which is the best fitting straight line through your data points.</b> Most commonly, the best fitting line is the line that minimizes errors. The equation for the regression line is what is used to make predictions for your dependent variable. 

<center><img src='https://s3.amazonaws.com/stackabuse/media/linear-regression-python-scikit-learn-1.png'></center>

***
## Things to Consider when using Linear Regression

***

### Feature Selection

Once we get a sense of the relationship between our variables, we need to make some decisions on which variables to include in our regression analyses. <b>We only want to include the variables that have some kind of relationship with your dependent variable(s) of interest.</b> This allows us to cut back on the empty "noise" in your dataset and focus on the variables that are meaningful. 

### Redundancy
Be careful with including multiple variables that show similar information. For example, if you have a variable "Hours of Study" which represents the total hours that a student studied, and you've also binned that data and created a new variable "Study Level" which groups students into "high, med, low" - you now have two variables in your dataset that show similar information, although one variable is a lot more specific. When you have this situation, you should elect to keep the variable that has the more detailed information (i.e. the specific hours of study). 

### Multiple Groups
Variables with a lot of different levels (i.e. State of Residence) don't do well in regression models. If you have a variable that has a large number of different groups, unless this variable is vital to your analyses, you should work to reduce the number of overall groups (i.e State can become Region), or leave the variable out of the analyses. It is possible to include these complicated variables, but it isn't always the easiest interpretation.

### Confounding Variables
When we preform regression analyses, we are interested in further understanding the relationship between the dependent and independent variable. Simply put, we are interested in investigating how the independent variable effects the dependent variable. However, it's rarely this simple and there are other factors that need to be understood and controlled. A confounding variable is a third variable that influences both the independent and dependent variable to some degree. It is important to acknowledge this third variable to ensure that the results of your analyses are valid. 

For example, you collect data on sunburns and ice cream consumption. You find that higher ice cream consumption is associated with a higher rate of sunburns. Does this mean that eating ice cream causes sunburns? Absolutely not, there are several other factors that can be attributed to this trend -- but most likely, the confounding variable is temperature. Hot temperatures cause people to eat more ice cream and result in people spending more time outdoors, which can result in more sunburns. Without accounting for the confounding variable(s), you may find relationships between variables that might not actually exist. To control for the potential effects of confounding variables, you simply have to include them in your regression model as another independent variable. 

When you include all of your assumed confounding variables in your regression model, you are controlling for the effects of all of them, and if you find there is still a relationship between a specific independent variable and your dependent variable, you will know that relationship isn't being influenced by any of these other factors.

In [19]:
## new library alert! ##

## Import the StatsModels library for our regression analyses

import statsmodels.formula.api as sm

### Creating a Linear Regression Model

* <b>result</b> is the name that we are assigning the regression formula
* <b>sm</b> is the shorthand for the linear regression model library
* <b>ols</b> is Ordinary Least Squares, the most common method of calculating the regression line
* the regression equation starts with the dependent variable on the left, followed by the independent variables
* independent variables are separated by "+"
* categorical variables must be in parentheses and annotated with a "C"
* data = is where you specify your dataset that you're pulling variables from
* <b>.fit()</b> function uses the predictive values to calculate the best linear regression line
* <b>.summary()</b> function will show the calculated values (slopes and y-intercept) for the linear regression formula

In [24]:
## create the regression model
result = sm.ols('grade ~ hours + exercise -1', data = df).fit()

## print the regression model summary
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                  grade   R-squared (uncentered):                   0.956
Model:                            OLS   Adj. R-squared (uncentered):              0.956
Method:                 Least Squares   F-statistic:                          2.160e+04
Date:                Fri, 17 Jun 2022   Prob (F-statistic):                        0.00
Time:                        01:11:34   Log-Likelihood:                         -8559.5
No. Observations:                2000   AIC:                                  1.712e+04
Df Residuals:                    1998   BIC:                                  1.713e+04
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
hours          4.9769      0.063     78.412      0.000       4.852       5.101
exercise       7.7302      0.224     34.521      0.000       7.291       8.169
==============================================================================
Omnibus:                        0.614   Durbin-Watson:                   1.852
Prob(Omnibus):                  0.736   Jarque-Bera (JB):                0.677
Skew:                          -0.011   Prob(JB):                        0.713
Kurtosis:                       2.913   Cond. No.                         7.11
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

# <font color=DODGERBLUE>Interpreting Linear Regression Result</font>
***
### Determining Model Fit
***
Linear regression calculates the regression line (equation) that minimizes the distance between the regression line and all the data points.  If our data points fall closely to the generated regression line, we consider the model to be a good fit. 

<img src='https://blog.minitab.com/hubfs/Imported_Blog_Media/residual_illustration-1.gif'>

But what does that mean non-graphically? Linear regression may not always be the right technique to use for the specific set of data. The fit of the model describes how well your variables explain the variance in the dependent variable. 

To assess the model fit, we look at the adjusted R-squared (Adj. R-squared). The Adj. R-squared is a statistical measure of how closely the data are to the fitted regression line. <B>The Adj. R-squared is the percentage of variation in the dependent variable that can be explained by all the independent variables included in the model.</B> For example, how much of the variation in student grades (i.e. grades ranging from 32-100) can be explained by hours of study, age, hours of exercise, gender, etc? Values range from 0 to 1; the higher the value, the better the fit. 

***
### Intercept
***
The y-intercept, or constant, is the value given to the dependent variable if all independent variables are equal to 0. For example, when all independent variables are equal to 0, the expected student grade is approximately 58. Don't worry too much about this interpretation, oftentimes this won't make sense - for example, if age equals 0, but it's important to know how to review this. 

***
### Coefficients (coef)
***
These values show how changes in the independent variable influence the dependent variable, and in what direction. 

#### <b> Interpreting Numeric Variable Coefficients </b>
When you are looking at numeric variables (i.e. age), each coefficient represents the numeric change in the dependent variable given a one-unit change in the independent variable. For example, for every one hour increase in study time (hours), grade increases by 1.9 points.

* <b>INTERCEPT</b>: when all other IV's are zero, expected grade is around 58

* <b>AGE</b>: for every one year increase in age, grade increases by 0.04 points (when controlling for hours of study, exercise, and gender)

* <b>EXERCISE</b>: for every one hour increase in exercise, grade increases by approximately 1 point (when controlling for age, hours of study, and gender)

* <b>HOURS</b>: for every one hour increase in study time, grade increases by 1.9 points (when controlling for age, hours of exercise, and gender)

If you have a variable that gives you a negative coefficient, the relationship moves in opposite directions. For example, say the coefficient for age was <b>- 0.0405</b>, you would interpret it in the following way: 

* <b>AGE</b>: for every one year increase in age, grade <u>decreases</u> by 0.04 points. 

#### <b> Interpreting Categorical Variable Coefficients </b>

When working with categorical variables, the model will automatically take one of the categories and use it as a reference (i.e. comparison category). This reference category will not show up in the listed coefficients, and that is how you will be able to identify which category is serving as the reference. In our model, gender is the only categorical variable, and can be interpreted: 

<b>GENDER</b>:
* Female is the reference category
* On average, Male students have a grade 0.44 points lower than Female students (when controlling for age, hours of study, and exercise) 

The grade is lower because the coefficient is negative in this model. If we have more than 2 categories for our categorical variable, we would still compare each level to the reference. For example, if we had a group of students who choose not to disclose their gender, the coefficient for the 'undisclosed' group would show up in our model with it's own coefficient and we would compare those results to the Female category. 

#### Standard Error (std err)

The standard error reflects the level of accuracy of the coefficients. The lower the value, the higher the level of accuracy.

***
### Statistical Significance 
***
#### p-value (P>|t|)
When trying to determine if the results we received are statistically important - we need to consider the p-value. The p-value is the probability that you will receive the same results solely by chance (aka there is no meaning behind the results). 

Because the p-value is representing the probability of random findings, we always want to minimize this value. If the p-value was .5 (50%), that would mean that 50% of the time the results we see are just by chance. The p-value reflects how confident we are in our results and requires strict interpretation. A commonly used cut-off is 0.05 (5% chance the results observed are by chance) or below. 
* If the p-value is less than or equal to 0.05, we can deem our results to be statistically significant. 
* If the p-value is greater than 0.05, our results are not statistically significant. 

***
### What do I do with non-significant variables?
***
Regression analyses are rarely a "one and done" situation. After you run your analyses, you should tweak your model (equation) and see if you can improve the model fit. A good place to start is by removing non-significant variables from our model. In this example, gender and age are not significant - let's try a model without them!

In [25]:
## Removing non-significant variables 

result2 = sm.ols('grade ~ hours + exercise', data = df).fit()

result2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  grade   R-squared:                       0.664
Model:                            OLS   Adj. R-squared:                  0.664
Method:                 Least Squares   F-statistic:                     1973.
Date:                Fri, 17 Jun 2022   Prob (F-statistic):               0.00
Time:                        01:22:00   Log-Likelihood:                -6300.8
No. Observations:                2000   AIC:                         1.261e+04
Df Residuals:                    1997   BIC:                         1.262e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     58.5316      0.447    130.828      0.000      57.654      59.409
hours          1.9162      0.031     61.575      0.000       1.855       1.977
exercise       0.9892      0.089     11.131      0.000       0.815       1.163
==============================================================================
Omnibus:                      318.721   Durbin-Watson:                   2.048
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2158.000
Skew:                          -0.564   Prob(JB):                         0.00
Kurtosis:                       7.962   Cond. No.                         43.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

# <font color=DODGERBLUE>Comparing Models and Making Predictions</font>

When comparting two models, the first thing you want to check is if there are any changes in the adjusted r-squared. In this example, the adj r-squared as not changed at all between the two models. The next item you want to check is the p-value of the remaining variables in the model. Did removing variables increase/decrease the p-values for the remaining variables?

You do not need to continue running your model until all your variables are significant. You should focus on maximizing your adj r-square, regardless of the significance of your variables. 

### What can I do with this information?

Now that you have a better understanding of how your variables interact, you are in a better place to describe your data. You now know that an increase of just one hour of studying will, on average, show a significant increase in a students final grade.  You can also make predictions about future grades...

### Making Predictions based on the Regression Results

Recall that a simple linear regression model is mathematically represented by the formula of a line: <b> y = mx + b</b>. When you are creating a multiple linear regression equation, the formula is similar - with some added features: <b> y = b + (m1 x X1) + (m2 x X2) + (m3 x X3) + ....etc. </b> where each "m x X" is representative of one of the coefficients in your model, and "b" represents the intercept. Once you have your regression output, you can plug these values in to make predictions on your dependent variable given specific values for your independent variables.

    grade(y) = intercept(b) + [hours of study coef(m1) x hours of study(x1)] + [hours of exercise coef(m2) x hours exercise(x2)]

    grade(y) = 58.5316 + [1.9162 x hours of study(x1)] + [0.9892 x hours exercise(x2)]

### What grade can we expect from a student that studied 8 hours and exercised 4.3 hours?

    grade(y) = 58.5316 + [1.9162 x hours of study(8)] + [0.9892 x hours exercise(4.3)]

    grade(y) = 58.5316 + [1.9162 x (8)] + [0.9892 x (4.3)]

    grade(y) = 58.5316 + [‭15.3296‬] + ‭[4.25356‬]

    grade = ‭78.11476

## <font color=SALMON>Ta Da! You can now predict the future!</font>

<LEFT><img src='https://i0.wp.com/www.learning-mind.com/wp-content/uploads/2019/10/psychic-spiritual-energy.jpg?resize=768%2C512&ssl=1'></LEFT>

## Making Predictions: the simple way!

What grade can we expect from a 16 year old female student that studied 8 hours and exercised 5.7 hours? 

In [26]:
## We can use the predict function (from statsmodel library) to predict the outcome given specific input
## reference the model with your function to reference the appropriate coef's!

# model_name.predict({'variable1_name':value1, 'variable2_name':value2, ...})

result.predict({
    'hours': 8, 
    'age': 16, 
    'exercise': 5.7, 
    'gender': "female"})

0    83.877386
dtype: float64

In [27]:
## What about another scenario?

result.predict({
    'hours': 14, 
    'age': 18, 
    'exercise': 12, 
    'gender': "male"})

0    162.439035
dtype: float64

# <font color=DODGERBLUE>Model 11 Exercises</font>

#### 1. Complete the code below to import the four libraries we've used most commonly.

In [259]:
import pandas as pd 
import numpy as np
import scipy.stats as stats
import statsmodels.formula.api as sm

#### 2. Import the "babies.csv" file and name it df. 

<b>Background Info</b>

    The Child Health and Development Studies considered all pregnancies between 1960 and 1967 among women in the Kaiser Foundation Health Plan in the San Francisco East Bay area. The goal is to model the weight of the infants (bwt, in ounces) using variables including length of pregnancy in days (gestation), mother's age in years (age), mother's height in inches (height), whether the child was the first born (parity), mother's pregnancy weight in pounds (weight), and whether the mother was a smoker (smoke).

<b>Variables</b>

    case - id number
    bwt - birthweight, in ounces
    gestation - length of gestation, in days
    parity - binary indicator for a first pregnancy (0=first pregnancy)
    age - mother's age in years
    height - mother's height in inches
    weight - mother's weight in pounds
    smoke - binary indicator for whether the mother smokes

In [260]:
df = pd.read_csv('babies.csv')

In [261]:
df

case  bwt  gestation  parity   age  height  weight  smoke
0        1  120      284.0       0  27.0    62.0   100.0    0.0
1        2  113      282.0       0  33.0    64.0   135.0    0.0
2        3  128      279.0       0  28.0    64.0   115.0    1.0
3        4  123        NaN       0  36.0    69.0   190.0    0.0
4        5  108      282.0       0  23.0    67.0   125.0    1.0
...    ...  ...        ...     ...   ...     ...     ...    ...
1231  1232  113      275.0       1  27.0    60.0   100.0    0.0
1232  1233  128      265.0       0  24.0    67.0   120.0    0.0
1233  1234  130      291.0       0  30.0    65.0   150.0    1.0
1234  1235  125      281.0       1  21.0    65.0   110.0    0.0
1235  1236  117      297.0       0  38.0    65.0   129.0    0.0

[1236 rows x 8 columns]

#### 3. Check the shape of the dataset. How many columns and rows are there?

In [262]:
df.shape

(1236, 8)

In [263]:
# there are 1236 rows and 8 columns in the dataset

#### 4. Check the first 10 rows and the last 10 rows. Drop the column "case". 

In [264]:
df.head(10)

case  bwt  gestation  parity   age  height  weight  smoke
0     1  120      284.0       0  27.0    62.0   100.0    0.0
1     2  113      282.0       0  33.0    64.0   135.0    0.0
2     3  128      279.0       0  28.0    64.0   115.0    1.0
3     4  123        NaN       0  36.0    69.0   190.0    0.0
4     5  108      282.0       0  23.0    67.0   125.0    1.0
5     6  136      286.0       0  25.0    62.0    93.0    0.0
6     7  138      244.0       0  33.0    62.0   178.0    0.0
7     8  132      245.0       0  23.0    65.0   140.0    0.0
8     9  120      289.0       0  25.0    62.0   125.0    0.0
9    10  143      299.0       0  30.0    66.0   136.0    1.0

In [265]:
df.tail(10)

case  bwt  gestation  parity   age  height  weight  smoke
1226  1227  109      244.0       1  21.0    63.0   102.0    1.0
1227  1228  103      278.0       0  30.0    60.0    87.0    1.0
1228  1229  118      276.0       0  34.0    64.0   116.0    0.0
1229  1230  127      290.0       0  27.0    65.0   121.0    0.0
1230  1231  132      270.0       0  27.0    65.0   126.0    0.0
1231  1232  113      275.0       1  27.0    60.0   100.0    0.0
1232  1233  128      265.0       0  24.0    67.0   120.0    0.0
1233  1234  130      291.0       0  30.0    65.0   150.0    1.0
1234  1235  125      281.0       1  21.0    65.0   110.0    0.0
1235  1236  117      297.0       0  38.0    65.0   129.0    0.0

In [266]:
df.drop(columns = 'case', inplace = True)

In [267]:
df

bwt  gestation  parity   age  height  weight  smoke
0     120      284.0       0  27.0    62.0   100.0    0.0
1     113      282.0       0  33.0    64.0   135.0    0.0
2     128      279.0       0  28.0    64.0   115.0    1.0
3     123        NaN       0  36.0    69.0   190.0    0.0
4     108      282.0       0  23.0    67.0   125.0    1.0
...   ...        ...     ...   ...     ...     ...    ...
1231  113      275.0       1  27.0    60.0   100.0    0.0
1232  128      265.0       0  24.0    67.0   120.0    0.0
1233  130      291.0       0  30.0    65.0   150.0    1.0
1234  125      281.0       1  21.0    65.0   110.0    0.0
1235  117      297.0       0  38.0    65.0   129.0    0.0

[1236 rows x 7 columns]

#### 5. Is there any missing data? Check!

In [268]:
df.isnull().sum()

bwt           0
gestation    13
parity        0
age           2
height       22
weight       36
smoke        10
dtype: int64

In [269]:
df.loc[df.isnull().any(axis = 1)]

bwt  gestation  parity   age  height  weight  smoke
3     123        NaN       0  36.0    69.0   190.0    0.0
39    128      282.0       0  31.0    65.0     NaN    0.0
42    138      302.0       0  26.0     NaN     NaN    1.0
85    129      274.0       0  29.0    71.0     NaN    1.0
89    114        NaN       0  24.0    67.0   113.0    1.0
...   ...        ...     ...   ...     ...     ...    ...
1044  117      323.0       0  26.0    62.0     NaN    0.0
1177  146      305.0       0  23.0     NaN     NaN    0.0
1190  104      275.0       0  24.0     NaN     NaN    0.0
1192  124        NaN       1  39.0    65.0   228.0    0.0
1215  152      267.0       0  28.0     NaN   119.0    1.0

[62 rows x 7 columns]

#### 6. The amount of missing data is small considering the size of our dataset. Drop all rows in the dataset that have missing data.

In [270]:
df.dropna(inplace = True)

In [271]:
df

bwt  gestation  parity   age  height  weight  smoke
0     120      284.0       0  27.0    62.0   100.0    0.0
1     113      282.0       0  33.0    64.0   135.0    0.0
2     128      279.0       0  28.0    64.0   115.0    1.0
4     108      282.0       0  23.0    67.0   125.0    1.0
5     136      286.0       0  25.0    62.0    93.0    0.0
...   ...        ...     ...   ...     ...     ...    ...
1231  113      275.0       1  27.0    60.0   100.0    0.0
1232  128      265.0       0  24.0    67.0   120.0    0.0
1233  130      291.0       0  30.0    65.0   150.0    1.0
1234  125      281.0       1  21.0    65.0   110.0    0.0
1235  117      297.0       0  38.0    65.0   129.0    0.0

[1174 rows x 7 columns]

#### 7. Check each of your numeric columns for outliers - pick one method and use it for all the columns. 

In [272]:
dfz = df.copy()

In [273]:
dfz

bwt  gestation  parity   age  height  weight  smoke
0     120      284.0       0  27.0    62.0   100.0    0.0
1     113      282.0       0  33.0    64.0   135.0    0.0
2     128      279.0       0  28.0    64.0   115.0    1.0
4     108      282.0       0  23.0    67.0   125.0    1.0
5     136      286.0       0  25.0    62.0    93.0    0.0
...   ...        ...     ...   ...     ...     ...    ...
1231  113      275.0       1  27.0    60.0   100.0    0.0
1232  128      265.0       0  24.0    67.0   120.0    0.0
1233  130      291.0       0  30.0    65.0   150.0    1.0
1234  125      281.0       1  21.0    65.0   110.0    0.0
1235  117      297.0       0  38.0    65.0   129.0    0.0

[1174 rows x 7 columns]

In [274]:
print(dfz.shape)

(1174, 7)


In [275]:
dfz['zscore_bwt'] = np.abs(stats.zscore(dfz['bwt']))
 

In [276]:
dfz.head()

bwt  gestation  parity   age  height  weight  smoke  zscore_bwt
0  120      284.0       0  27.0    62.0   100.0    0.0    0.029337
1  113      282.0       0  33.0    64.0   135.0    0.0    0.352741
2  128      279.0       0  28.0    64.0   115.0    1.0    0.465998
4  108      282.0       0  23.0    67.0   125.0    1.0    0.625654
5  136      286.0       0  25.0    62.0    93.0    0.0    0.902658

In [277]:
z_outliers = dfz.loc[dfz['zscore_bwt'] > 3].index


In [278]:
print(z_outliers)

Int64Index([632, 829, 912, 978, 1139], dtype='int64')


In [279]:
dfz.iloc[[ 632, 829, 912, 978, 1139 ]]

bwt  gestation  parity   age  height  weight  smoke  zscore_bwt
676   100      275.0       0  26.0    60.0   115.0    0.0    1.062315
881   128      272.0       1  18.0    67.0   109.0    0.0    0.465998
967   119      273.0       0  35.0    65.0   125.0    1.0    0.025246
1035   65      237.0       0  31.0    67.0   130.0    0.0    2.972705
1200   97      255.0       1  22.0    63.0   107.0    1.0    1.226062

In [280]:
dfz = dfz.drop(z_outliers)
 

In [281]:
print(dfz.shape)

(1169, 8)


In [282]:
# to find outliers for gestation

In [283]:
dfz['zscore_gestation'] = np.abs(stats.zscore(dfz['gestation']))

In [284]:
dfz.head()

bwt  gestation  parity   age  height  weight  smoke  zscore_bwt  \
0  120      284.0       0  27.0    62.0   100.0    0.0    0.029337   
1  113      282.0       0  33.0    64.0   135.0    0.0    0.352741   
2  128      279.0       0  28.0    64.0   115.0    1.0    0.465998   
4  108      282.0       0  23.0    67.0   125.0    1.0    0.625654   
5  136      286.0       0  25.0    62.0    93.0    0.0    0.902658   

   zscore_gestation  
0          0.301151  
1          0.173990  
2          0.016752  
4          0.173990  
5          0.428312

In [285]:
z_outliers = dfz.loc[dfz['zscore_gestation'] < -3].index

In [286]:
print(z_outliers)

Int64Index([], dtype='int64')


In [287]:
dfz.drop(z_outliers)

bwt  gestation  parity   age  height  weight  smoke  zscore_bwt  \
0     120      284.0       0  27.0    62.0   100.0    0.0    0.029337   
1     113      282.0       0  33.0    64.0   135.0    0.0    0.352741   
2     128      279.0       0  28.0    64.0   115.0    1.0    0.465998   
4     108      282.0       0  23.0    67.0   125.0    1.0    0.625654   
5     136      286.0       0  25.0    62.0    93.0    0.0    0.902658   
...   ...        ...     ...   ...     ...     ...    ...         ...   
1231  113      275.0       1  27.0    60.0   100.0    0.0    0.352741   
1232  128      265.0       0  24.0    67.0   120.0    0.0    0.465998   
1233  130      291.0       0  30.0    65.0   150.0    1.0    0.575163   
1234  125      281.0       1  21.0    65.0   110.0    0.0    0.302250   
1235  117      297.0       0  38.0    65.0   129.0    0.0    0.134411   

      zscore_gestation  
0             0.301151  
1             0.173990  
2             0.016752  
4             0.173990  
5             0.428312  
...                ...  
1231          0.271074  
1232          0.906879  
1233          0.746214  
1234          0.110409  
1235          1.127697  

[1169 rows x 9 columns]

In [288]:
print(dfz.shape)

(1169, 9)


In [309]:
dfz['zscore_age'] = np.abs(stats.zscore(dfz['age']))

In [310]:
dfz.head()

bwt  gestation  parity   age  height  weight  smoke  zscore_bwt  \
0  120      284.0       0  27.0    62.0   100.0    0.0    0.029337   
1  113      282.0       0  33.0    64.0   135.0    0.0    0.352741   
2  128      279.0       0  28.0    64.0   115.0    1.0    0.465998   
4  108      282.0       0  23.0    67.0   125.0    1.0    0.625654   
5  136      286.0       0  25.0    62.0    93.0    0.0    0.902658   

   zscore_gestation  zscore_age  zscore_height  
0          0.301151    0.035023       0.841336  
1          0.173990    1.001431       0.029637  
2          0.016752    0.137719       0.029637  
4          0.173990    0.725993       1.187913  
5          0.428312    0.380508       0.841336

In [311]:
z_outliers = dfz.loc[dfz['zscore_age'] > 3 ].index

In [312]:
print(z_outliers)

Int64Index([], dtype='int64')


In [313]:
dfz.iloc[[1070]]

bwt  gestation  parity   age  height  weight  smoke  zscore_bwt  \
1136  122      285.0       0  31.0    62.0   102.0    1.0    0.138502   

      zscore_gestation  zscore_age  zscore_height  
1136          0.364731    0.655946       0.841336

In [314]:
dfz = dfz.drop(z_outliers)

In [315]:
print(dfz.shape)

(1164, 11)


In [316]:
dfz['zscore_height'] = np.abs(stats.zscore(dfz['height']))

In [317]:
dfz.head()

bwt  gestation  parity   age  height  weight  smoke  zscore_bwt  \
0  120      284.0       0  27.0    62.0   100.0    0.0    0.029337   
1  113      282.0       0  33.0    64.0   135.0    0.0    0.352741   
2  128      279.0       0  28.0    64.0   115.0    1.0    0.465998   
4  108      282.0       0  23.0    67.0   125.0    1.0    0.625654   
5  136      286.0       0  25.0    62.0    93.0    0.0    0.902658   

   zscore_gestation  zscore_age  zscore_height  
0          0.301151    0.035023       0.841336  
1          0.173990    1.001431       0.029637  
2          0.016752    0.137719       0.029637  
4          0.173990    0.725993       1.187913  
5          0.428312    0.380508       0.841336

In [318]:
z_outliers = dfz.loc[dfz['zscore_height'] > 3 ].index

In [319]:
print(z_outliers)

Int64Index([], dtype='int64')


In [ ]:
# no outliers for height

In [320]:
print(dfz.shape)

(1164, 11)


In [321]:
dfz['zscore_weight'] = np.abs(stats.zscore(dfz['weight']))

In [322]:
dfz.head()

bwt  gestation  parity   age  height  weight  smoke  zscore_bwt  \
0  120      284.0       0  27.0    62.0   100.0    0.0    0.029337   
1  113      282.0       0  33.0    64.0   135.0    0.0    0.352741   
2  128      279.0       0  28.0    64.0   115.0    1.0    0.465998   
4  108      282.0       0  23.0    67.0   125.0    1.0    0.625654   
5  136      286.0       0  25.0    62.0    93.0    0.0    0.902658   

   zscore_gestation  zscore_age  zscore_height  zscore_weight  
0          0.301151    0.035023       0.841336       1.377494  
1          0.173990    1.001431       0.029637       0.317435  
2          0.016752    0.137719       0.029637       0.651096  
4          0.173990    0.725993       1.187913       0.166830  
5          0.428312    0.380508       0.841336       1.716480

In [323]:
z_outliers = dfz.loc[dfz['zscore_weight'] > 3].index

In [324]:
print(z_outliers)

Int64Index([88, 117, 149, 181, 183, 426, 522, 563, 608, 622, 723, 849, 858,
            865, 924, 1148],
           dtype='int64')


In [325]:
dfz.iloc[[88, 117, 149, 181, 183, 426, 522, 563, 608, 622, 723, 849, 858, 865, 924, 1148]]

bwt  gestation  parity   age  height  weight  smoke  zscore_bwt  \
95    135      282.0       0  22.0    64.0   100.0    0.0    0.848076   
128   109      272.0       0  35.0    66.0   154.0    0.0    0.571071   
163   135      296.0       0  30.0    63.0   123.0    0.0    0.848076   
199   137      291.0       0  34.0    61.0   110.0    0.0    0.957241   
201   142      275.0       0  25.0    64.0   132.0    0.0    1.230154   
461   112      281.0       1  23.0    61.0   150.0    0.0    0.407324   
562   161      302.0       1  22.0    70.0   170.0    1.0    2.267223   
604   109      302.0       0  24.0    64.0   110.0    0.0    0.571071   
654   146      306.0       0  38.0    63.0   112.0    0.0    1.448484   
669   116      275.0       1  20.0    68.0   145.0    0.0    0.188993   
776   126      285.0       0  24.0    64.0   140.0    0.0    0.356832   
907   127      272.0       0  20.0    64.0   130.0    1.0    0.411415   
917   128      279.0       0  27.0    66.0   135.0    0.0    0.465998   
925   163      289.0       1  25.0    64.0   126.0    1.0    2.376388   
987   118      276.0       0  29.0    62.0   130.0    1.0    0.079828   
1220  135      284.0       1  19.0    60.0    95.0    0.0    0.848076   

      zscore_gestation  zscore_age  zscore_height  zscore_weight  
95            0.173990    0.898735       0.029637       1.377494  
128           0.461815    1.346915       0.782063       1.237540  
163           1.064117    0.483204       0.435487       0.263683  
199           0.746214    1.174173       1.247186       0.893229  
201           0.271074    0.380508       0.029637       0.172156  
461           0.110409    0.725993       1.247186       1.043834  
562           1.445600    0.898735       2.405462       2.012365  
604           1.445600    0.553250       0.029637       0.893229  
654           1.699922    1.865142       0.435487       0.796376  
669           0.271074    1.244220       1.593762       0.801701  
776           0.364731    0.553250       0.029637       0.559568  
907           0.461815    1.244220       0.029637       0.075302  
917           0.016752    0.035023       0.782063       0.317435  
925           0.619053    0.380508       0.029637       0.118404  
987           0.207493    0.310461       0.841336       0.075302  
1220          0.301151    1.416962       1.653036       1.619627

In [326]:
dfz = dfz.drop(z_outliers)

In [327]:
print(dfz.shape)

(1148, 12)


#### 9. Let's model birthweight based on the characteristics of the mother. But first... 

We want to easily distinguish between the numeric and categorical variables. Replace the values 0/1 in the "parity" and "smoke" column with meaningful labels (i.e. smokes, doesn't smoke).

In [876]:
df['parity'].replace([0,1],['N','Y'],inplace =  True)

In [877]:
df

case  bwt  gestation parity   age  height  weight  smoke
0        1  120      284.0      N  27.0    62.0   100.0    0.0
1        2  113      282.0      N  33.0    64.0   135.0    0.0
2        3  128      279.0      N  28.0    64.0   115.0    1.0
3        4  123        NaN      N  36.0    69.0   190.0    0.0
4        5  108      282.0      N  23.0    67.0   125.0    1.0
...    ...  ...        ...    ...   ...     ...     ...    ...
1231  1232  113      275.0      Y  27.0    60.0   100.0    0.0
1232  1233  128      265.0      N  24.0    67.0   120.0    0.0
1233  1234  130      291.0      N  30.0    65.0   150.0    1.0
1234  1235  125      281.0      Y  21.0    65.0   110.0    0.0
1235  1236  117      297.0      N  38.0    65.0   129.0    0.0

[1236 rows x 8 columns]

In [878]:
df['smoke'].replace([0.0, 1.0],['doesn\'t smoke','smoke'], inplace = True)

In [879]:
df

case  bwt  gestation parity   age  height  weight          smoke
0        1  120      284.0      N  27.0    62.0   100.0  doesn't smoke
1        2  113      282.0      N  33.0    64.0   135.0  doesn't smoke
2        3  128      279.0      N  28.0    64.0   115.0          smoke
3        4  123        NaN      N  36.0    69.0   190.0  doesn't smoke
4        5  108      282.0      N  23.0    67.0   125.0          smoke
...    ...  ...        ...    ...   ...     ...     ...            ...
1231  1232  113      275.0      Y  27.0    60.0   100.0  doesn't smoke
1232  1233  128      265.0      N  24.0    67.0   120.0  doesn't smoke
1233  1234  130      291.0      N  30.0    65.0   150.0          smoke
1234  1235  125      281.0      Y  21.0    65.0   110.0  doesn't smoke
1235  1236  117      297.0      N  38.0    65.0   129.0  doesn't smoke

[1236 rows x 8 columns]

In [880]:
result = sm.ols('bwt ~ age + gestation + parity + height  + weight + smoke ', data = df).fit()

In [881]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    bwt   R-squared:                       0.258
Model:                            OLS   Adj. R-squared:                  0.254
Method:                 Least Squares   F-statistic:                     67.61
Date:                Fri, 17 Jun 2022   Prob (F-statistic):           2.87e-72
Time:                        21:39:39   Log-Likelihood:                -4904.7
No. Observations:                1174   AIC:                             9823.
Df Residuals:                    1167   BIC:                             9859.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept        -80.4109     14.347     -5.605      0.000    -108.559     -52.263
parity[T.Y]       -3.3272      1.129     -2.947      0.003      -5.542      -1.112
smoke[T.smoke]    -8.4007      0.954     -8.807      0.000     -10.272      -6.529
age               -0.0090      0.086     -0.104      0.917      -0.177       0.159
gestation          0.4440      0.029     15.258      0.000       0.387       0.501
height             1.1540      0.205      5.629      0.000       0.752       1.556
weight             0.0502      0.025      1.987      0.047       0.001       0.100
==============================================================================
Omnibus:                        6.574   Durbin-Watson:                   2.061
Prob(Omnibus):                  0.037   Jarque-Bera (JB):                8.427
Skew:                           0.029   Prob(JB):                       0.0148
Kurtosis:                       3.411   Cond. No.                     9.80e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.8e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### 10. Run a correlation matrix with your dataset. Which variables are correlated with birthweight? 

Describe the strength of the correlation between all the numeric variables and birthweight. 

In [882]:
df.corr()

case       bwt  gestation       age    height    weight
case       1.000000 -0.057643   0.012473  0.007451 -0.038316 -0.055327
bwt       -0.057643  1.000000   0.407854  0.033649  0.198519  0.154038
gestation  0.012473  0.407854   1.000000 -0.056269  0.063631  0.022366
age        0.007451  0.033649  -0.056269  1.000000  0.003051  0.158275
height    -0.038316  0.198519   0.063631  0.003051  1.000000  0.435745
weight    -0.055327  0.154038   0.022366  0.158275  0.435745  1.000000

In [883]:

# gestation has almost moderate  positive relationship with birthweight
# age , height and weight have weak positive relatioship with birthweight


#### 11. Determine the relationship between birthweight and the categorical variables: parity and smoke. 

Use the groupby function to determine if there are any differences between birthweight and the different groups.  Does it seem like there is a relationship between these variables and birthweight?

In [884]:
df['bwt'].groupby(df['parity']).mean()

parity
N    120.068404
Y    118.139683
Name: bwt, dtype: float64

In [885]:
df['bwt'].groupby(df['smoke']).mean()

smoke
doesn't smoke    123.047170
smoke            114.109504
Name: bwt, dtype: float64

In [886]:
# there is some relationship between these variables and birthweight

#### 12. Let's construct your regression model. Firstly, which variables do you plan to include in your model, and why? 

In the space below, write your justification for why you are including each variable. 

In [887]:
# gestation, height, weight,  and smoke are included in my model. 

In [888]:
# My justification to include these variables is: they are independent vriables that affect the dependent vriable birthweight.
# regarding gestation, I think the more the number of weeks the infant stayed in mother womb might increase the birthweight
# at the same time  height and weight of the mother also contributes for the increase or decrease of the birhtweight, so the more the height and weight of the mother it might lead to  the more the birthweight. 
# scientifically, smoking of the mother during pregnancy is related with health issues which latter affect the birthweight.  

#### 13. Construct your regression model and print the summary. 

Write out your full interpretation of the regression results. If you are not happy with the results, tweak your model and run it again. 

In [889]:
result = sm.ols('bwt ~ age + gestation + parity + height  + weight + smoke ', data = df).fit()

In [890]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    bwt   R-squared:                       0.258
Model:                            OLS   Adj. R-squared:                  0.254
Method:                 Least Squares   F-statistic:                     67.61
Date:                Fri, 17 Jun 2022   Prob (F-statistic):           2.87e-72
Time:                        21:39:53   Log-Likelihood:                -4904.7
No. Observations:                1174   AIC:                             9823.
Df Residuals:                    1167   BIC:                             9859.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept        -80.4109     14.347     -5.605      0.000    -108.559     -52.263
parity[T.Y]       -3.3272      1.129     -2.947      0.003      -5.542      -1.112
smoke[T.smoke]    -8.4007      0.954     -8.807      0.000     -10.272      -6.529
age               -0.0090      0.086     -0.104      0.917      -0.177       0.159
gestation          0.4440      0.029     15.258      0.000       0.387       0.501
height             1.1540      0.205      5.629      0.000       0.752       1.556
weight             0.0502      0.025      1.987      0.047       0.001       0.100
==============================================================================
Omnibus:                        6.574   Durbin-Watson:                   2.061
Prob(Omnibus):                  0.037   Jarque-Bera (JB):                8.427
Skew:                           0.029   Prob(JB):                       0.0148
Kurtosis:                       3.411   Cond. No.                     9.80e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.8e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [891]:
result = sm.ols('bwt ~ + gestation + parity + height  + weight + smoke ', data = df).fit()

In [892]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    bwt   R-squared:                       0.258
Model:                            OLS   Adj. R-squared:                  0.255
Method:                 Least Squares   F-statistic:                     81.26
Date:                Fri, 17 Jun 2022   Prob (F-statistic):           2.68e-73
Time:                        21:39:55   Log-Likelihood:                -4908.5
No. Observations:                1175   AIC:                             9829.
Df Residuals:                    1169   BIC:                             9859.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
Intercept        -80.6357     14.037     -5.745      0.000    -108.176     -53.096
parity[T.Y]       -3.2917      1.062     -3.099      0.002      -5.376      -1.207
smoke[T.smoke]    -8.3834      0.950     -8.823      0.000     -10.248      -6.519
gestation          0.4440      0.029     15.280      0.000       0.387       0.501
height             1.1538      0.205      5.639      0.000       0.752       1.555
weight             0.0500      0.025      1.999      0.046       0.001       0.099
==============================================================================
Omnibus:                        6.606   Durbin-Watson:                   2.061
Prob(Omnibus):                  0.037   Jarque-Bera (JB):                8.489
Skew:                           0.028   Prob(JB):                       0.0143
Kurtosis:                       3.413   Cond. No.                     9.56e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 9.56e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [893]:
# let see how each independent variable affect the dependent variables just by looking the Adj.R-squared value
# our Adj.R-squared value is 0.254 which is a small value and the p-value for age is 0.917 which is greater than the limit 0.05 so age is not significant
# by removing ages our Adj.R-squared is increase to 0.255 and all our p-value indicates less than 0.05 which is the standard limit for good relationship
# this indicates that all independent variables except age have relationships with the dependent variable bwt.

#### 14. Create three scenarios (i.e. make up specific values) and predict the birthweight given these factors. Use the information in the model summary to make these predictions. 

In [895]:
result.predict({
      'gestation': 260,
      'height': 60,
      'age': 40,   
       'weight': 130,
       'parity': 'N',
        'smoke': 'smoke'})

0    102.14722
dtype: float64

In [896]:
result.predict({
    'gestation': 200,
    'height': 75,
    'age': 35,
    'weight': 145,
    'parity': 'Y',
    'smoke': 'smoke'})

0    90.271921
dtype: float64

In [899]:
result.predict({
    'gestation': 275,
    'height': 70,
    'age': 25,
    'weight': 160,
    'parity': 'Y',
    'smoke': 'doesn\'t smoke'})

0    126.936551
dtype: float64